src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+        try {
+            final JWSBuilder jwsBuilder = new JWSBuilder();
+            final Oauth2 oauthRequest = new Oauth2(mAuthRequest, mWebRequestHandler, jwsBuilder);
+            result = oauthRequest.refreshToken(refreshToken);
+            if (result != null && StringExtensions.IsNullOrBlank(result.getRefreshToken())) {
+                Logger.v(TAG, "Refresh token is not returned or empty");
+                result.setRefreshToken(refreshToken);
+            }
+        } catch (final IOException | AuthenticationException exc) {
+            // Server side error or similar
+            Logger.e(TAG, "Error in refresh token for request:" + mAuthRequest.getLogInfo(),
+                    ExceptionExtensions.getExceptionMessage(exc), ADALError.AUTH_FAILED_NO_TOKEN,
+                    exc);
+
+            AuthenticationException authException = new AuthenticationException(
+                    ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
+                    exc);
+            throw authException;
+        }
+
+        return result;
+    }
+    
+    /**
+     * For testing purpose, inject the dependency. 
+     */
+    void setWebRequestHandler(final IWebRequestHandler webRequestHandler) {
+        this.mWebRequestHandler = webRequestHandler;
+    }
+    
+    /**
+     * Attempt to get new access token with regular RT. 
+     */
+    private AuthenticationResult tryRT() throws AuthenticationException {
+        final String user = getUser();
+        final TokenCacheItem regularRTItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), user);
+        
+        if (regularRTItem == null) {
+            Logger.v(TAG, "Regular token cache entry does not exist, try with MRRT.");
+            return tryMRRT(); 
+        }
+        
+        // When MRRT is returned, we store separate entries for both regular RT entry and MRRT entry, 
+        // we should look for MRRT entry if the token in regular RT entry is marked as MRRT.
+        // However, the current cache implementation never mark the token stored in regular RT entry
+        // as MRRT. To support the backward compatibility and improve cache lookup, when successfully
+        // retrieved regular RT entry token and if the mrrt flag is false, check the existence of MRRT.
+        if (regularRTItem.getIsMultiResourceRefreshToken() || isMRRTExisted()) {
+            Logger.v(TAG, "Regular token cache entry exists but it's marked as MRRT, try with MRRT.");
+            return tryMRRT();
+        }
+        
+        Logger.v(TAG, "Send request to use regular RT for new AT.");
+        return acquireTokenWithCachedItem(regularRTItem);
+    }
+    
+    /**
+     * Attempt to get new access token with MRRT. 
+     */
+    private AuthenticationResult tryMRRT() throws AuthenticationException {
+
+        if (mMrrtTokenCacheItem == null) {
```

Comment:
put a comment on we check for null here
add a boolean flag for checking if already visit the method